File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
  private PriorityQueue<ListNode, int> q = new();
  public ListNode MergeKLists(ListNode[] lists) {
    if (lists.Length == 0) return null;
    foreach (ListNode list in lists) {
      if (list != null)
        q.Enqueue(list, list.val);
    }
    ListNode dummy = new ListNode();
    ListNode p = dummy;
    while (q.Count > 0) {
      p.next = q.Dequeue();
      p = p.next;
      if (p.next != null) {
        q.Enqueue(p.next, p.next.val);
      }
    }
    return dummy.next;
  }
}

File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.go
func removeEmpty(lists []*ListNode) []*ListNode {
  slow := 0
  for _, v := range lists {
    if v != nil {
      lists[slow] = v
      slow++
    }
  }
  return lists[:slow]
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeKLists(lists []*ListNode) *ListNode {
  lists = removeEmpty(lists)
  switch len(lists) {
  case 0: return nil
    case 1: return lists[0]
  default:
    ans := &ListNode{}
    p := ans
    for len(lists) > 1 {
      sm := 0
      for i := 1; i < len(lists); i++ {
        if lists[i].Val < lists[sm].Val {
          sm = i
        }
      }
      p.Next = lists[sm]
      p = p.Next
      if lists[sm].Next == nil {
        lists[len(lists)-1], lists[sm] = lists[sm], lists[len(lists)-1]
        lists = lists[:len(lists)-1]
      } else {
        lists[sm] = lists[sm].Next
      }
    }
    p.Next = lists[0]
    return ans.Next
  }
}